File: bench/ruby/pfib.rb
# Parallel fib: compute fib(N) in 100 tasks at once, summed. Ruby's GVL means
# threads give no CPU speedup, so the idiomatic way to use cores is forking
# worker processes (like Python's multiprocessing). A pool sized to the core
# count splits the 100 tasks; each child computes its share and pipes back a
# subtotal. Checksum = 100*fib(N).
require "etc"

n = (ENV["BENCH_N"] || "30").to_i
tasks = 100

def fib(n)
  n < 2 ? n : fib(n - 1) + fib(n - 2)
end

workers = [tasks, Etc.nprocessors].min
readers = []

workers.times do |w|
  # strided assignment: how many of the `tasks` fibs this worker owns
  count = 0
  i = w
  while i < tasks
    count += 1
    i += workers
  end

  r, wio = IO.pipe
  fork do
    r.close
    sub = 0
    count.times { sub += fib(n) }
    wio.write(sub.to_s)
    wio.close
  end
  wio.close
  readers << r
end

total = 0
readers.each do |r|
  total += r.read.to_i
  r.close
end
Process.waitall

puts total

File: bench/node/pfib.js
// Parallel fib: compute fib(N) in 100 tasks at once, summed. Node is
// single-threaded, so using cores for CPU work means worker_threads — a pool
// sized to the core count, with the 100 tasks split across the workers.
// Checksum = 100*fib(N).
const { Worker, isMainThread, parentPort, workerData } = require("worker_threads");
const os = require("os");

const N = parseInt(process.env.BENCH_N || "31", 10);
const TASKS = 100;

function fib(n) {
  return n < 2 ? n : fib(n - 1) + fib(n - 2);
}

if (isMainThread) {
  const P = Math.min(TASKS, os.cpus().length);
  let pending = P;
  let total = 0;
  for (let w = 0; w < P; w++) {
    // strided assignment: how many of the 100 tasks this worker owns
    let count = 0;
    for (let i = w; i < TASKS; i += P) count++;
    const worker = new Worker(__filename, { workerData: { count, n: N } });
    worker.on("message", (sub) => {
      total += sub;
      if (--pending === 0) console.log(total);
    });
    worker.on("error", (e) => {
      throw e;
    });
  }
} else {
  const { count, n } = workerData;
  let sub = 0;
  for (let i = 0; i < count; i++) sub += fib(n);
  parentPort.postMessage(sub);
}
